main/src/main/scala/sbt/Cross.scala

```diff
@@ -323,11 +329,11 @@ object Cross {
           .map(BuildRef.apply)
           .map(proj => proj -> crossVersions(extracted, proj))
       } else {
-        val binaryVersion = CrossVersion.binaryScalaVersion(version)
+        val binaryVersion = crossCompatibleVersion(version)
```

Comment:
Thanks for the PR. Looking at this change, my first thought was that if there's some compatibility logic regarding Scala versions, we should centralize that in librarymanagement objects like `CrossVersion.*`. And then I realized that `CrossVersion.binaryScalaVersion` proxies `CrossVersionUtil.binaryScalaVersion`, which does in fact centralizes the Scala version compatibility logic.
https://github.com/sbt/librarymanagement/blob/v1.7.0-M1/core/src/main/scala/sbt/internal/librarymanagement/cross/CrossVersionUtil.scala#L61-L91

Excerpt related to Scala 3:

```scala
  private[sbt] def binaryScala3Version(full: String): String = full match {
    case ReleaseV(maj, _, _, _)                                                  => maj
    case NonReleaseV_n(maj, min, patch, _) if min.toLong > 0 || patch.toLong > 0 => maj
    case BinCompatV(maj, min, patch, stageOrNull, _) =>
      val stage = if (stageOrNull != null) stageOrNull else ""
      binaryScala3Version(s"$maj.$min.$patch$stage")
    case _ => full
  }

  def binaryScalaVersion(full: String): String = {
    if (ScalaArtifacts.isScala3(full)) binaryScala3Version(full)
    else
      binaryVersionWithApi(full, TransitionScalaVersion)(scalaApiVersion) // Scala 2 binary version
  }
```

Aside: I'm not a fan of the use of the term "binary" in this context because it literally can have 2 different meanings:
1. Of two things, as in first two segments `x.y` of `x.y.z`.
2. Of machine-code things, like "binary compatibility" of JAR files.
In this case `CrossVersionUtil.binaryScalaVersion` returns ABI version (application binary interface version), which for Scala 3 should be `"3"`.

In other words, the current implementation performs a kind of a fuzzy search on `++<v1> <command>` where `v1` is compared against `crossScalaVersions` using `binaryScalaVersion` (ABI version) as the test.

One thing that might be related is https://github.com/sbt/sbt/discussions/6893, where I (unknowingly) proposed a more formal fuzzy search `3.*` e.g. `++3.* test`

If we can get https://github.com/sbt/sbt/pull/6894 in for 1.7.0, then I wonder if we should remove the fuzziness out of non-`*` versions like `2.12.16` and `3.1.2`. Then I think the implementation would be more straightforward to code and explain.

Reply:
@eed3si9n I usually don't think of `++` as some kind of "string-based" fuzzy search on Scala versions, and I think the operator becomes much less useful if it behaves like that.

I usually apply it in the context of having different subprojects supporting independent sets of versions, and when I write something like `++<v1> <command>`, I read it as "run \<command\> in all projects that can be compiled with \<v1\>". With the introduction of Scala 3 this doesn't seem to be a simple string search anymore. If I want to express things like the original situation of #6915, I could write `++3.1.* <command>`. However, once Scala 3.2 comes out I'd have to run both `++3.1.* <command>; ++3.2.* <command>` to achieve the behavior described above, and so on.

My preference is for the operator to encode implicitly the relationship of "project targeting Scala version X can run with Scala version Y". Scala 3 seems to have made that relation unidirectional - a project targeting Scala 3.0 can use Scala 3.1, but not the other way around. But this is still about binary compatibility, not about version regexes.

Reply:
> My preference is for the operator to encode implicitly the relationship of "project targeting Scala version X can run with Scala version Y".

Understood. If that's the case, should we should implement a match function that encodes that in https://github.com/sbt/librarymanagement/blob/v1.7.0-M1/core/src/main/scala/sbt/librarymanagement/CrossVersionExtra.scala and document it. If we go by "X can run with Y" test, shouldn't that allow subproject with 3.0.0 to be included when the user types `++3.2.0 test`?

Reply:
Sure, I can send a PR to librarymanagement with this new function 👍 It seems useful even outside this use case.

> If we go by "X can run with Y" test, shouldn't that allow subproject with 3.0.0 to be included when the user types ++3.2.0 test?

Yes, that's my expectation. That's a real use case for libraries that want to specify Scala 3.0.x as the target version for publishing (to maximize compatibility) but still want to ensure their code works against the latest Scala 3 version as per [this recommendation](https://www.scala-lang.org/blog/2021/10/21/scala-3.1.0-released.html).

Reply:
That may work for testing, but we'd still have issues regarding publishing though.

```
++3.2.0 publishSigned
```

would now try to publish core subproject twice.

Reply:
That's true, but running these two commands will result in problems even treating versions as strict, since the binary artifact for them is the same (ending with `_3` suffix). Library maintainers are never expected to publish the same artifact using two different Scala 3 versions.

That said, I see how it might be confusing for users. If you think these edge cases might be confusing for users, what do you think of adding a new modifier for backwards binary compatibility, like the bang operator in `++ <version>!`?

Reply:
Since picking binary compatible has been the behavior of `++`, changing to _backward_ binary compatible is still an improvement. I think we just have to document its limitations with regards to publishing etc now that we have a better picture of what's going on.

Reply:
Sounds good! Opened https://github.com/sbt/librarymanagement/pull/400.